### PR TITLE
Added cache for primary index

### LINKED ISSUE #71992: Cache for primary index
**Use case**

In some extremely large workloads (trillions of rows) primary index may take significant amount of memory: tens or hundreds of GBs. Instead we can use cache for primary similar to mark cache instead of always storing index for parts.

**Describe the solution you'd like**

Use a cache similar to mark cache. If lazy load and primary index cache are enabled store index in that cache instead of storing it in data part.

**Describe alternatives you've considered**

Lazy load of index, but the problem is that index is never evicted and once primary index was analyzed for some query the index will be stored in memory until restart or `SYSTEM UNLOAD PRIMARY KEY` query.

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added cache for primary index of `MergeTree` tables (can be enabled by table setting `use_primary_key_cache`). If lazy load and cache are enabled for primary index, it will be loaded to cache on demand (similar to mark cache) instead of keeping it in memory forever. Added prewarm of primary index on inserts/mergs/fetches of data parts and on restarts of table (can be enabled by setting `prewarm_primary_key_cache`).

Resolves #71992.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 985255912e7e805d27ba9048701b77dc82443f42 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72102/985255912e7e805d27ba9048701b77dc82443f42/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/72102/985255912e7e805d27ba9048701b77dc82443f42/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by CurtizJ** [comment]
Fuzzer: #72265.

**Comment by alesapin** [comment]
0_o

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp [hunk 1]
```diff
@@ -365,46 +371,61 @@ IMergeTreeDataPart::~IMergeTreeDataPart()
...
+    if (auto index_cache = storage.getPrimaryIndexCache())
+        return loadIndexToCache(*index_cache);
+
+    index = loadIndex();
     return index;
 }
 
+IMergeTreeDataPart::IndexPtr IMergeTreeDataPart::loadIndexToCache(PrimaryIndexCache & index_cache) const
+{
+    auto key = PrimaryIndexCache::hash(getDataPartStorage().getFullPath());
```
Comment (alesapin) [inline src/Storages/MergeTree/IMergeTreeDataPart.cpp:390]: Maybe to a comment somewhere, what we use as a key for cache, but it's clear.

**Review by alesapin — APPROVED** [review]
LGTM

**Comment by clickgapai** [comment]
Hi @CurtizJ @alesapin — while reviewing this PR I found the following:

- **Bug (P2):** `getRequiredAccessForDDLOnCluster` checks wrong permission for `PREWARM_PRIMARY_INDEX_CACHE` — https://github.com/ClickHouse/ClickHouse/pull/104802

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**